### DISCUSSION TITLE
Free chat

### DISCUSSION THREAD

**Discussion by panno8M** [godot-nim/.github#3] [General]
Chit chat, or small talk that doesn't require starting a new discussion

**Comment by panno8M** [comment]
@insomniacUNDERSCORElemon 

If you don't mind, could you make an icon for our organization?
Like godot mascot wearing a crown.
The style of expression, such as effects, etc., is up to you.

Please think on it,

**Reply by insomniacUNDERSCORElemon** [reply]
I was already working on a crown svg because of the info you gave me in the other issue. Though it doesn't actually work well for an icon (on my 1080p screen, node icons are 16x16... maybe it could be 32x without changing the size of other stuff, but I don't know if that's an option) so the pixel art one is a bit better there.

I don't know if it's the best placement, but here's icon.svg with the crown added:

![icon](https://github.com/user-attachments/assets/b5e75077-0901-471f-82a8-43902c1dc084)

(just remember the copyright notice for the godot icon)

I had a gradient on the front part too, but it looked a little out of place (I left it on the back so the color wasn't as ugly, and a bit of depth).

I could perhaps do a different crown design (or even an entirely new Godot head)... I've thought a little on different interpretations but haven't tried anything.

EDIT: Not sure if it's just my browser, but for me the crown renders incorrectly in a new tab. Here's a no-gradient version:

![crown_icon_nogradient](https://github.com/user-attachments/assets/9aa4de57-2bc5-4075-aaa3-f8c9a9049184)

Nah that's not it, I don't know what it's doing.

**Reply by panno8M** [reply]
Nice. I said I'd leave the style to you, but yes, pixel art is fascinating. Can it be more pixel'ish?

**Reply by panno8M** [reply]
cutey?
![pppixelate](https://github.com/user-attachments/assets/77dffcba-044f-4c7a-9698-58a8195f4b37)

**Reply by insomniacUNDERSCORElemon** [reply]
> Nice. I said I'd leave the style to you, but yes, pixel art is fascinating. Can it be more pixel'ish?

I haven't done much pixel art in a while (I used to have a resource pack, but even then the pixel art wasn't great), but here's a rusty/rushed attempt (40x40):

![crowned_godot](https://github.com/user-attachments/assets/58057ae5-b0c0-4cee-b2aa-4ee37e39c497) Scaled 2x: ![crowned_godot_upscale](https://github.com/user-attachments/assets/4468d7b6-b14e-4ac3-9e44-0e799f5598ea)
Scaled 4x: 
![crowned_godot_upscale](https://github.com/user-attachments/assets/b8cde085-d211-4747-9b42-8a39b4b4210b)

EDIT: I should've made the teeth wider

**Reply by panno8M** [reply]
Cool! May I use it for our project?
Please let me know if you have any specifications for the credits. If not otherwise, I will use `CC-by insomniac_lemon, godot-nim`.

**Reply by insomniacUNDERSCORElemon** [reply]
Sure. Here's the .kra file (Krita document) with some changes (zipped because it wouldn't let me upload .kra):
[crown_godot.kra.zip](https://github.com/user-attachments/files/16402250/crown_godot.kra.zip)

Re-exported, non-scaled:

![crowned_godot](https://github.com/user-attachments/assets/b26e2106-8c7a-4717-8643-09401a31d4f0)

Note particularly if you upscale it a lot: if you try to optimize the image, the color profile might mess up.

**Reply by insomniacUNDERSCORElemon** [reply]
You should upscale (nearest neighbor). It's currently a bit blurry on the overview page. I didn't do it (other than a preview) because the need will always be different, depending on display size.

@panno8M

**Reply by panno8M** [reply]
@insomniacUNDERSCORElemon 
[project overview](https://github.com/godot-nim) is updated again. Please check it

**Reply by insomniacUNDERSCORElemon** [reply]
@panno8M I've updated the layered image a bit to create an icon for the library (pre-coronation) part:

[crown_godot_v2.kra.zip](https://github.com/user-attachments/files/16405045/crown_godot_v2.kra.zip)

It's otherwise the same as before, the only difference is an extra layer only visible when the crown layer is turned off (as it is by default in the version above).

native scale: 
![library_godot](https://github.com/user-attachments/assets/d9946b94-e7b4-4135-a6c2-b2ea4d2c16cb)

4x: 
![library_godot_4x](https://github.com/user-attachments/assets/af9a2dba-ccbc-4b2a-8795-a0c30c0d575b)

I also just noticed that the icon view (not overview, but when in a repository) of this design isn't so great because it's re-scaled from 40x40 to 48x48. You could manually add padding to get a slightly-smaller yet crisper icon.

Probably not such a big deal, something someone better could re-pixel later (it might end up being me, but this is why I prefer polygons).

EDIT: Also another reply (on `self: Window`) at the bottom of the page.

**Reply by panno8M** [reply]
As a matter of fact, github icons are per host, we can't change them per repository. But it's very nice art and I'll use it for demos and tutorials.

**Comment by insomniacUNDERSCORElemon** [comment]
Two of the same:

`self.getNode("grid/a/outlabel").Label.setText("hmm")`

`self.getNode("grid/a/outlabel").Label.text = "hmm"`

Neither of these give an error, neither works and if they're included they prevent the signal from firing more than once. Am I doing something wrong?

Note that if I try the entire original line:

 `self.getNode("grid/a/outlabel").Label.text = convert_from_decimal(self.getNode("grid/inbox").LineEdit.text)`

It actually does spit out an error about a type mismatch, `string` (my code) vs `String` (from gdLineEdit.nim)

EDIT: Actually there are capitalized and non-capitalized versions in that one file. Inconsistency or non-standard string type for Godot?

**Reply by panno8M** [reply]
Typecasting in godot requires calling a dedicated method, so you must use `(node as Label)` or `node.castTo(Label)`.

**Comment by panno8M** [comment]
@insomniacUNDERSCORElemon 
I have created this issue godot-nim/docs#1 and would appreciate your help when you have time.

**Reply by insomniacUNDERSCORElemon** [reply]
As I've mentioned, I am not really a programmer so if I stumble it is often just because of that. Though I'd say lacking config.nims (assuming that causes errors) and casting trouble are not really docs issues but are in need of dedicated error messages at least. On a similar note, when the gdsync pragma is used but not needed it will spit out an unrelated error message.

The classes thing we've already discussed being helped by an output message (or alternate grouping in the add node prompt, putting classes in their own group rather than by the node they inherit).

I am still working on my demo project, I'll tell you if I run into more issues. Not really liking the castTo thing. A simple change sure, but the first line I'm trying to convert over having a type error still:

`self.getNode(out_id).castTo(Label).text = convert_from_decimal(self.getNode(in_id).castTo(LineEdit).text)`

![feeling_it](https://github.com/user-attachments/assets/7ed267a4-cae3-42cf-829e-9238fd046fee)

EDIT: Are you sure there isn't an issue with you using `String` instead of `string`? Because I can re-recreate that mismatch just by trying to godot.echo a `String`.

**Reply by panno8M** [reply]
> EDIT: Are you sure there isn't an issue with you using String instead of string? Because I can re-recreate that mismatch just by trying to godot.echo a String.

Nim has a function called converter that, if defined, will automatically convert when there is a type mismatch.
Since converters from string are defined for String, StringName, and NodePath, they basically work without any problems.
https://github.com/godot-nim/godot/blob/1d3e7db9bca5303f673384665a9b73b1488619e7/src/godot/conversions.nim#L6-L8

I will do some research on the echo thing.

**Reply by panno8M** [reply]
> I am still working on my demo project, I'll tell you if I run into more issues. Not really liking the castTo thing. A simple change sure, but the first line I'm trying to convert over having a type error still:

Could you please paste the error message?

**Reply by insomniacUNDERSCORElemon** [reply]
> Could you please paste the error message?
```
Error: type mismatch
Expression: convert_from_decimal(getText(castTo(getNode(self, in_id), LineEdit)))
  [1] getText(castTo(getNode(self, in_id), LineEdit)): String

Expected one of (first mismatch at [position]):
[1] proc convert_from_decimal(input_decimal: string): string
```

I haven't used capital-S string anywhere, the main reason for me to not use snake_case is if I think it won't work (which is more often than I'd like). gdLineEdit.nim and gdLabel.nim both use `String` (although I didn't see an inconsistency between them, oddly).

EDIT: Actually `String` doesn't cause a problem with echo, it causes issues with getNode:

```
Error: type mismatch
Expression: getNode(self, out_id)
  [1] self: MainClass
  [2] out_id: String

Expected one of (first mismatch at [position]):
[2] proc getNode(self: Node; path: NodePath): Node
```

**Reply by panno8M** [reply]
probably,

`self.getNode(in_id).castTo(LineEdit).text` is type of godot.String
And there are no converters godot.String -> system.string
So, `self.getNode(out_id).castTo(Label).text = convert_from_decimal($self.getNode(in_id).castTo(LineEdit).text)` may resolve it.

**Reply by panno8M** [reply]
Similarly, there is no String -> NodePath converter; consider keeping out_id as NodePath.

**Reply by insomniacUNDERSCORElemon** [reply]
> Similarly, there is no String -> NodePath converter; consider keeping out_id as NodePath.

`out_id` was a string that I changed to String to cause that error as an example. The actual issue was the generic nim function (`convert_from_decimal()`) not wanting the string output by Godot, but the error did not properly reflect that.

Anyways that was the last major hurdle for the number converter. The long function pretty much worked without much issue (after find+replace in another document to do the footwork). I'm actually having more issue now that I'm moving on to moving a polygon. I'm getting a core dump.

```
Traceback (most recent call last)
/home/username/.nimble/pkgs2/godotgen-4.2.2-6fdc96a164750b10074e5bcdd7cd4078da9447de/godotgen/classes/gdNode.nim(13) process
/home/username/.nimble/pkgs2/godotgen-4.2.2-6fdc96a164750b10074e5bcdd7cd4078da9447de/godotgen/classes/gdNode.nim(11) process
/run/media/username/fast_storage/godot4_new/nim/decbinhex4/src/polygon_class.nim(17) process
/home/username/.nimble/pkgs2/godotgen-4.2.2-6fdc96a164750b10074e5bcdd7cd4078da9447de/godotgen/classes/gdInput.nim(77) isActionPressed
SIGSEGV: Illegal storage access. (Attempt to read from nil?)

```

The file:

```

import godot, godotgen/classes/gdSceneTree
import godotgen/classes / [gdInput, gdNode2D, gdPolygon2D]

type PolygonClass* = ref object of Polygon2D
  Input:Input
  InputMap:InputMap

method ready(self:PolygonClass) {.gdsync.} = 
  if self.getTree.editedSceneRoot != nil: return
  print("present")

#tank movement
method process(self:PolygonClass, delta: float64) {.gdsync.} =
  if self.getTree.editedSceneRoot != nil: return #I don't know if this was needed but I added it when it crashed before even running
  if self.Input.isActionPressed("ui_up"): self.moveLocalY(-1.0'f64)
  elif self.Input.isActionPressed("ui_down"): self.moveLocalY(+0.85'f64)
  if self.Input.isActionPressed("ui_right"): self.setRotation(self.getRotation + 0.01'f64)
  elif self.Input.isActionPressed("ui_left"): self.setRotation(self.getRotation - 0.01'f64)

```

I mean I know the crash is related to the process method, and probably related to types.

I had the same error related to a checkbox in the main class, but I somehow fixed it without noticing what exactly it was.

EDIT: Also I should say this error puts nothing in the Godot output (even in the cases where it doesn't crash the editor with it), only to the terminal.

**Reply by panno8M** [reply]
Input and InputMap are singletons.
In this code, only a variable is defined, and its value is nil.

You can resolve this by calling `self.Input = singleton Input` in the ready method.
Instead of `singleton Input`, you can use `/Input`; the two have the same meaning.

```diff
method ready(self:PolygonClass) {.gdsync.} = 
  if self.getTree.editedSceneRoot != nil: return
  print("present")
+ self.Input = /Input
+ self.InputMap = /InputMap
```

**Reply by insomniacUNDERSCORElemon** [reply]
I'm trying to get window min/max size working now, running into the issue with `self: Window`. Neither 3.X nim era nor the Godot 4 docs seem to need this context. Does this need to be done in its own class? Or something like bolting on `type WindowClass* = ref object of Window` into any class? I'm not sure what'd be after that, unless it'd just need to run in a `proc some_name(self: WindowClass)` with `self.gdWindow.setMinSize(vector2i(325, 200))`

EDIT: No.

Also this would be really elegant:

```
type WindowClass* = ref object of Window
  minSize = vector2i(325, 200)
```
but despite no error, doesn't seem to work either.

**Reply by panno8M** [reply]
You can get current window with `Node.getWindow()`.
Maybe what you'd like to do is something like:

```nim
method ready*(self: Main) {.gdsync.} =
  self.window = self.getWindow()
  self.window.minSize = vector2i(325, 200)
```

reference: [Setting window size - Help - Godot Forum](https://forum.godotengine.org/t/setting-window-size/47572)

**Reply by insomniacUNDERSCORElemon** [reply]
> Maybe what you'd like to do is something like:

Worked after I added `window:Window` to the ref object, thank you.

**Comment by insomniacUNDERSCORElemon** [comment]
Not sure if you've been following with it, but [4.3 just got a stable release](https://godotengine.org/releases/4.3/). Some big changes, some to GDextension too (I think you can explicitly set code as a tool script rather than needing to stop that every time).

**Reply by insomniacUNDERSCORElemon** [reply]
I forgot to say it'd be good to update to this so users' systems can be updated without needing to ignore the `godot` package or altering gdextgen. More relevant as time goes on.

**Reply by panno8M** [reply]
The supported version has been updated to 4.3. (I made a mistake and forgot to raise the PR.)
I would like to do something about this only, because when I use an older version of Godot, it throws an error and quits right after startup even if you are not using any of the new features from 4.3.

**Reply by insomniacUNDERSCORElemon** [reply]
[4.4 released](https://godotengine.org/releases/4.4/).

EDIT: Also whoops, I just saw issue 165 so looks like you saw this long before I did

(sorry to drop in just to say that, I do have some ideas I have been thinking about but have not felt right enough to start them yet)

**Reply by panno8M** [reply]
Thanks for the notice. And don't care. Glad to see you are still interested in the project.

**Comment by insomniacUNDERSCORElemon** [comment]
Am I doing something wrong here?

Trying to get mouse input for a 3D camera.

```
method unhandledInput(self:Player, event:gdref InputEvent): void {.gdsync.} =
  if Engine.isEditorHint: return

  if event is InputEventMouseMotion: print "YES"
  if event is not InputEventMouseMotion: print event
```

The `is not` line is always called, even for mouse motion (the line above it is not called, more obvious if the `is not` line is commented out).

Meta question: is there any way to mark-as-resolved (or otherwise hide) questions in this thread? Or will you just need to make new threads as-needed?

**Reply by panno8M** [reply]
For now, we can only write as follows:

```nim
  let keyevent = event as gdref InputEventKey
  if keyevent[].isNil: # if failed to typecast:
    # bla bla bla...
```

(I am planning to be able to write using Nim's `of`, but that is a difficult task because of too heavy compile time due to.)

> Meta question: is there any way to mark-as-resolved (or otherwise hide) questions in this thread? Or will you just need to make new threads as-needed?

I think it would be a good idea to start a new Q&A thread for each question.
We have opened a new Discussion in gdext-nim repository, so please use it instead.
https://github.com/godot-nim/gdext-nim/discussions

![Peek 2024-11-27 15-36](https://github.com/user-attachments/assets/c9ac47f1-ae1b-451c-bbd1-f857aeb2698f)

**Reply by insomniacUNDERSCORElemon** [reply]
>  if keyevent[].isNil

So, this is an issue or I can still use the mouse motion with the event being nil?

> that is a difficult task because of too heavy compile time due to.

I don't really need the explanation here, just pointing out it's not there.

> I think it would be a good idea to start a new Q&A thread for each question.

Well yes. I was just commenting on the "Free chat" idea. I was hoping that tools existed to make it better... even having the default sort set to 'new' would help.

-------------------------------------------

> We have opened a new Discussion in gdext-nim repository, so please use it instead.

You might want to be more clear what each discussion board is for. I assume the new one is specifically for actual code, where the original might be for meta (or perhaps general gamedev) topics?

> tell us about what you do with this community

You should re-word this.

I would probably cut that whole line (starting with `To get started`) and go with something like `Introduce yourself, your development interests, and anything you have done with the project thus far!`. Admittedly I am not sure if the meaning/tone is in-line with your original wording, particularly relating to community involvement.

**Reply by panno8M** [reply]
## InputEvent Q&A

Oh, sorry, I should have used InputEventMouseMotion instead of InputEventKey for the example.
A complete example follows:

```nim
  let
    mousemotion = event as gdref InputEventMouseMotion
    key = event as gdref InputEventKey
  if mousemotion[] != nil:
    print "mouse relative: ", mousemotion[].screen_relative
  elif key[] != nil:
    print "key pressed: ", key[].pressed
  else:
    print event[]
```

If Godot fails to typecast (i.e., event is not an InputEventMouseMotion), `as` returns a nil object.

**Reply by panno8M** [reply]
## gdext-nim Discussion board

That's the default content that github suggests. I'll fix it later to fit this project.
